lib/chai/core/assertions.js

```diff
@@ -988,9 +988,33 @@ module.exports = function (chai, _) {
 
   function assertInstanceOf (constructor, msg) {
     if (msg) flag(this, 'message', msg);
+
+    var target = flag(this, 'object')
+    var validInstanceOfTarget = constructor === Object(constructor) && (
+        typeof constructor === 'function' ||
+        (typeof Symbol !== 'undefined' && typeof Symbol.hasInstance !== 'undefined' &&
+         typeof constructor[Symbol.hasInstance] === 'function')
```

Comment:
I guess the question is: Is there an implicit understanding that Chai, in the process of performing an `instanceof` assertion, will only trigger `@@hasInstance` once? Or is it allowed to trigger it as much as it wants?

The benefit of not worrying about calling it multiple times is that the code is less complex. The cost is that to test an object with an observable `@@hasInstance`, you'll have to manually use JavaScript's `instanceof` and then assert on the result of that, rather than using Chai's builtin `instanceof` assertion.

My gut feeling is that having an observable `@@hasInstance` is such an extreme edge case that it's not worth Chai jumping through any hoops to accommodate, especially given the easy workaround of invoking `instanceof` manually and asserting on the result. But it doesn't add too much complexity to accommodate it either, so I won't protest if we want to do it.

Anyone have strong feelings one way or another?